development_guidelines.md

```diff
@@ -0,0 +1,415 @@
+The html page shows the summary of the coverage data collected during the 
+execution of the tests.
+
+**Note:**  
+Make sure that all the tests are executed successfully because, otherwise,
+the coverage value does not reflect the correct coverage of your tests.
+
+## Specification
+If new properties are added to the system or if old ones have to be changed, 
+the specification of the system needs to be updated. 
+The specification helps each one of us to quickly gain an understanding what 
+and what the system cannot do.  
+Currently, our requirements are situated in a file called `spec.dox`. 
+To unify and simplify the creation of requirements, and the linking of
+requirements against tests, we introduced new doxygen commands which are
+explained in the following sub-sections.
+
+**Please note:**  
+The specification of a package is generated while generating the 
+documentation. For more information on how to generate the documentation for
+a package see section:
+[How-to generate documentation for a package](#how-to-generate-documentation-for-a-package)
+
+### How-to define and write a requirement
+To define a requirement the following doxygen command is used:
+```
+@definereq{my_requirements_name}
+Here comes the specification of my super cool new feature I'm adding.
+```
+**Some remarks how to formulate a requirement:**  
+In order, to keep the requirement simple and understandable, use short and
+plain sentences. If it difficult to deduce why the system needs to fulfill
+a certain requirement, please shortly state the reason. Also, do not forget to
+define how the system behaves if something goes wrong 
+(e.g. errors, etc.) or for edge cases (e.g. lowest/highest possible
+number, etc.)!  
+
+### How-to link a requirement against a test
+To state that a test checks a certain requirement or part of a requirement,
+the following doxygen command is used:
+```
+@tests{my_requirement_name, A one sentence long short description of the test (without comma!)}
+```
+**Note:**
+- A requirement is allowed to be tested by more than one test. In other words,
+a requirement can be linked against as many tests as needed.
+- A test is allowed to test more than on requirement. In other words, you are
+allowed to use as many `@tests` commands in the test description as needed.
+
+## Documentation
+To document the code, [Doxygen](http://www.doxygen.nl/) is used for C++ code
+and [Sphinx](https://www.sphinx-doc.org/en/master/index.html) for
+Python code.
+
+### How-to generate documentation for a package
+To generate the documentation for a package consisting of C++ or/and 
+Python code, a tool named [rosdoc_lite](http://wiki.ros.org/rosdoc_lite)
+is used:
+- How to install the tool see the ROS wiki:
+[rosdoc_lite Installation](http://wiki.ros.org/rosdoc_lite#Installation). 
+- How to use rosdoc_lite see the ROS wiki:
+ [Generate document locally](http://wiki.ros.org/rosdoc_lite#Generate_document_locally).
+
+## Testing
+
+### How-to document tests
```

Comment:
Now that we gained some experience on this I would really like to re-discuss this step by step test documentation. Up for it?

Reply:
I get your points by I figure that all of these points apply not only to test code but for production code as well.
While one can argue against some difference in rules for tests (e.g. DAMP vs DRY (see [link](https://stackoverflow.com/a/11837973)).

However when it comes to documentation imho Clean Code - Chapter 4,
, https://twitter.com/unclebobmartin/status/870311898545258497?lang=en seems pretty compelling.
There are reasons for comments in unittests (see https://blog.cleancoder.com/uncle-bob/2017/02/23/NecessaryComments.html) however they should originate from a lack of expressing the "why" not the "what". ( [https://blog.codinghorror.com/code-tells-you-how-comments-tell-you-why/](https://blog.codinghorror.com/code-tells-you-how-comments-tell-you-why/) )

> In this case, the test code has to explain itself. However, this approach does only work if I'm disciplined enough to always ensure that my test are easy to read and understand. But what happens for example if somebody refactors the test and does not ensure that the readability is maintained (This also happened before.)?

The documentation scheme does not help against a lack of discipline here. Let's say for the sake of the argument the code is written in a fast rather "as long as it works" fashion. Is it reasonable to assume that the same person would suddenly start to write a total expressiv, correct and well-thought comment above the questionable code?
Same holds for refactorings as well if not even more.

>Also: How do we define the term "readability" or "understandability" when it comes to test code?

I see no reason why this should be different from production code. Surely these are rather subjective terms and since these code qualities are completely irrelevant for the machine (where compilability is enough :wink:) since they address other programmers (including yourself in a few weeks :neckbeard:)

>Furthermore, this approach misses a short summary (one sentence) explaining what the test intents to do. This information is something that even very "good code" cannot express that easily.

If it is possible the name should indicate summarise "what".
(I also find [https://osherove.com/blog/2005/4/3/naming-standards-for-unit-tests.html](https://osherove.com/blog/2005/4/3/naming-standards-for-unit-tests.html) interesting)

>Even if the programmer knows the documentation strategy: Who guarantees that the programmer does update the documentation?

Maybe my mistake was that I only expressed what I disliked. So what I don't want is another "documentation strategy" but rather a precise usage of comments only at points where the reasoning cannot be expressed in any other way. (see again https://blog.cleancoder.com/uncle-bob/2017/02/23/NecessaryComments.html)

>Provide documentation scheme:
This is basically the current approach. Like @agutenkunst mentioned there are some issues with this scheme. Looking at the alternative approaches, however, we see that the issues of this approach are not so different from the other approaches: What happens if somebody refactors a test? Who guarantees that the documentation is updated?

Could you provide some references to alternativ approaches or to our current approach?

>I don't think that the current documentation scheme itself is the problem but rather our inability to automatically check the actuality of the documentation and the readability of the code.

This is something which will ~never~ only be possible by AI in a few years :stuck_out_tongue_winking_eye:
Tbh and excuse me if I am wrong: But did we maybe by introducing the current scheme try to create something measurable? (We could also check the scheme in the CI - just kidding - please don't)

>We can of course reduce the documentation overhead in order to reduce the amount of work one has to do if a test is refactored.

Totally up for it.

>But again: How do we define readability and understandability when it comes to testing? And: Who checks that the readability and understandability are maintained or exist in the first place?

I know this sounds naiv but: This should be checked by the reviewer.
Also this writing code with high expressiv and readable code is not something that comes for free by just writing enough lines for a few years. I think this needs to be learned explicitly and in exchange with other programmers reading the code one wrote. Thus reviews also focussing on this seem ideal for this.

To summarise what I would desire:
- Don't make the current scheme mandatory or even remove it from the guidelines
- Focus on the readability/expressiveness of the (test-)code itself
  - Encourage programmers to engage in group discussions if they are not sure if their code has these qualities
- Focus on creating useful testhelpers which allow to abstract code lines not relevant in detail and provide domain specific Macros (e.g. EXPECT_POSITION)
 
@ct2034 @leunMar @martiniil your opinion?

Reply:
I totally agree with keeping comments as close to / within the test (keeping-life-simple rather than "discipline" ;-)) and getting rid of the "headline" steps in the future.

Reply:
@agutenkunst  Although, I think this is a good and useful discussion, I suggested closing this PR without modifications and then opening a new PR whose sole purpose is to continue this discussion.

Reply:
How about removing the controversial part and opening a PR for integrating it?

Reply:
As you wish, sir. -> Done.